cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/BrokerSetAwareGoal.java

```diff
@@ -0,0 +1,380 @@
+                .maybeAddSelectionFunc(ReplicaSortFunctionFactory.selectReplicasBasedOnExcludedTopics(excludedTopics),
+                                       !excludedTopics.isEmpty())
+                .trackSortedReplicasFor(replicaSortName(this, tracksReplicasInReverseOrder, tracksOnlyLeaderReplicas), clusterModel);
+
+        BrokerSetDataStore brokerSetDataStore = _balancingConstraint.getBrokerSetDataStore();
+
+        _brokerSetDataHelper = new BrokerSetDataHelper(clusterModel, brokerSetDataStore);
+    }
+
+    /**
+     * Update goal state.
+     * Sanity check: After completion of balancing / self-healing, confirm that replicas of each topic reside fully in a
+     * broker set.
+     *
+     * @param clusterModel        The state of the cluster.
+     * @param optimizationOptions Options to take into account during optimization.
+     */
+    @Override
+    protected void updateGoalState(ClusterModel clusterModel, OptimizationOptions optimizationOptions)
+            throws OptimizationFailureException {
+        // Sanity check: No self-healing eligible replica should remain at a dead broker/disk.
+        GoalUtils.ensureNoOfflineReplicas(clusterModel, name());
+        // Sanity check: No replica should be moved to a broker, which used to host any replica of the same partition on its broken disk.
+        GoalUtils.ensureReplicasMoveOffBrokersWithBadDisks(clusterModel, name());
+        // Sanity check to confirm that the final distribution is broker set aware.
+        ensureBrokerSetAware(clusterModel, optimizationOptions);
+        if (_provisionResponse.status() != ProvisionStatus.OVER_PROVISIONED) {
+            _provisionResponse = new ProvisionResponse(ProvisionStatus.RIGHT_SIZED);
+        }
+        finish();
+    }
+
+    private void ensureBrokerSetAware(ClusterModel clusterModel, OptimizationOptions optimizationOptions)
+            throws OptimizationFailureException {
+        // Sanity check to confirm that the final distribution is brokerSet aware.
+        Set<String> excludedTopics = optimizationOptions.excludedTopics();
+        Map<String, Set<Broker>> brokersByBrokerSet = _brokerSetDataHelper.getBrokersByBrokerSet();
+
+        // Sanity check to check if brokersByBrokerSet map is not empty
+        if (brokersByBrokerSet == null || brokersByBrokerSet.isEmpty()) {
+            throw new OptimizationFailureException(String.format("[%s] broker set to brokers mapping cannot be empty.",
+                                                                 name()));
+        }
+
+        // Sanity check to check if all brokers in data store do not match all brokers in cluster model
+        Set<Broker> allBrokersInDataStore = brokersByBrokerSet.values()
+                .stream()
+                .flatMap(brokers -> brokers.stream())
+                .collect(Collectors.toSet());
+
+        if (!allBrokersInDataStore.equals(clusterModel.brokers())) {
```

Comment:
My understanding is that clustermodel.brokers() would have all brokers alive or dead. clustermodel has other functions like aliveBrokers() that would give alive brokers information. Correct me if I am wrong and if it would ever happen that brokers() results in not all brokers.

Reply:
I think you are right that brokers return all brokers regardless of being alive or dead. @efeg to confirm.

Reply:
Does that make sense to extract l214 to l224 to another function that's only called in initGoalState? I feel like this check only needs to be done before the optimization, as brokersByBrokerSet should not change under your implementation. 

However if we consider in the future, that the brokersByBrokerSet is a dynamic value, then it's fine to leave it here.

Reply:
We're also interested in what do you think would be a good way to handle broker adding/removing. With the current implementation, adding/removing brokers will inevitably failed the goal, as this if condition will break.

Reply:
Can this case ever happen? Shouldn't this (i.e. equality of `allBrokersInDataStore` and `clusterModel.brokers()`) already be guaranteed during construction at an earlier stage?

>> My understanding is that clustermodel.brokers() would have all brokers alive or dead. clustermodel has other functions like aliveBrokers() that would give alive brokers information. Correct me if I am wrong and if it would ever happen that brokers() results in not all brokers.
> 
> I think you are right that brokers return all brokers regardless of being alive or dead. @efeg to confirm.

This is correct.